src/main/java/org/checkerframework/specimin/AnnotationParameterTypesVisitor.java

```diff
@@ -0,0 +1,400 @@
+        // Replace with Class<?> to prevent compile-time errors
+        String type = "Class<?>";
+        if (decl.getType().isArrayType()) {
+          type += "[]";
+        }
+        decl.setType(type);
+      } else {
+        try {
+          ResolvedType resolved = decl.getType().resolve();
+          if (resolved.isArray()) {
+            resolved = resolved.asArrayType().getComponentType();
+          }
+          if (resolved.isReferenceType()) {
+            usedClass.add(resolved.asReferenceType().getQualifiedName());
+          }
+        } catch (UnsolvedSymbolException ex) {
+          // TODO: retrigger synthetic type generation
+          return super.visit(decl, p);
+        }
+      }
+      Optional<Expression> defaultValue = decl.getDefaultValue();
+      if (defaultValue.isPresent()) {
+        Set<String> usedClassByCurrentAnnotation = new HashSet<>();
+        Set<String> usedMembersByCurrentAnnotation = new HashSet<>();
+        boolean resolvable =
+            handleAnnotationValue(
+                defaultValue.get(), usedClassByCurrentAnnotation, usedMembersByCurrentAnnotation);
+
+        if (resolvable) {
+          classesToAdd.addAll(usedClassByCurrentAnnotation);
+          usedMembers.addAll(usedMembersByCurrentAnnotation);
+        }
+      }
+    }
+    return super.visit(decl, p);
+  }
+
+  @Override
+  public Visitable visit(MarkerAnnotationExpr anno, Void p) {
+    Node parent = findClosestParentMemberOrClassLike(anno);
+
+    if (isTargetOrUsed(parent)) {
+      handleAnnotation(anno);
+    }
+    return super.visit(anno, p);
+  }
+
+  @Override
+  public Visitable visit(SingleMemberAnnotationExpr anno, Void p) {
+    Node parent = findClosestParentMemberOrClassLike(anno);
+
+    if (isTargetOrUsed(parent)) {
+      handleAnnotation(anno);
+    }
+    return super.visit(anno, p);
+  }
+
+  @Override
+  public Visitable visit(NormalAnnotationExpr anno, Void p) {
+    Node parent = findClosestParentMemberOrClassLike(anno);
+
+    if (isTargetOrUsed(parent)) {
+      handleAnnotation(anno);
+    }
+    return super.visit(anno, p);
+  }
+
+  /**
+   * Finds the closest method, field, or class-like declaration (enums, annos)
+   *
+   * @param node The node to find the parent for
+   * @return the Node of the closest member or class declaration
+   */
+  public Node findClosestParentMemberOrClassLike(Node node) {
+    Node parent = node.getParentNode().orElseThrow();
+    while (!(parent instanceof ClassOrInterfaceDeclaration
+        || parent instanceof EnumDeclaration
+        || parent instanceof AnnotationDeclaration
+        || parent instanceof ConstructorDeclaration
+        || parent instanceof MethodDeclaration
+        || parent instanceof FieldDeclaration)) {
+      parent = parent.getParentNode().orElseThrow();
+    }
+    return parent;
+  }
+
+  /**
+   * Determines if the given Node is a target/used method or class. Should be used following
+   * findClosestParentMemberOrClassLike().
```

Comment:
Also, this method seems like it would be useful in some of the other `Visitor` classes. Maybe we should make a visitor superclass that contains this method and some of the fields that many of the visitors have?

(I'm not expecting you to do that in this PR, but I'd appreciate it if you could add a TODO comment to this method.)

Reply:
> I can see that that is the use case you wrote it for, but I think this method could be used otherwise without danger, no?

Yep, you're right. Sorry that I didn't clarify in the method, I just meant that it is intended to be used afterwards. Since it's not dangerous I removed the comment as it is unnecessary.

> (I'm not expecting you to do that in this PR, but I'd appreciate it if you could add a TODO comment to this method.)
Added👍